CMake/ExternalData.cmake

```diff
@@ -0,0 +1,1202 @@
+associated file options.  For example, the argument
+``DATA{MyDataDir/,REGEX:.*}`` will pass the full path to a ``MyDataDir``
+directory on the command line and ensure that the directory contains
+files corresponding to every file or content link in the ``MyDataDir``
+source directory.  In order to match associated files in subdirectories,
+specify a ``RECURSE:`` option, e.g. ``DATA{MyDataDir/,RECURSE:,REGEX:.*}``.
+
+Hash Algorithms
+^^^^^^^^^^^^^^^
+
+The following hash algorithms are supported::
+
+ %(algo)     <ext>     Description
+ -------     -----     -----------
+ MD5         .md5      Message-Digest Algorithm 5, RFC 1321
+ SHA1        .sha1     US Secure Hash Algorithm 1, RFC 3174
+ SHA224      .sha224   US Secure Hash Algorithms, RFC 4634
+ SHA256      .sha256   US Secure Hash Algorithms, RFC 4634
+ SHA384      .sha384   US Secure Hash Algorithms, RFC 4634
+ SHA512      .sha512   US Secure Hash Algorithms, RFC 4634
+ SHA3_224    .sha3-224 Keccak SHA-3
+ SHA3_256    .sha3-256 Keccak SHA-3
+ SHA3_384    .sha3-384 Keccak SHA-3
+ SHA3_512    .sha3-512 Keccak SHA-3
+ CID         .cid      Interplanetary Filesystem (IPFS) Content Identifier (CID)
+
+Note that the hashes are used only for unique data identification and
+download verification.
+
+.. _`ExternalData Custom Fetch Scripts`:
+
+Custom Fetch Scripts
+^^^^^^^^^^^^^^^^^^^^
+
+When a data file must be fetched from one of the URL templates
+specified in the ``ExternalData_URL_TEMPLATES`` variable, it is
+normally downloaded using the :command:`file(DOWNLOAD)` command.
+One may specify usage of a custom fetch script by using a URL
+template of the form ``ExternalDataCustomScript://<key>/<loc>``.
+The ``<key>`` must be a C identifier, and the ``<loc>`` must
+contain the ``%(algo)`` and ``%(hash)`` placeholders.
+A variable corresponding to the key, ``ExternalData_CUSTOM_SCRIPT_<key>``,
+must be set to the full path to a ``.cmake`` script file.  The script
+will be included to perform the actual fetch, and provided with
+the following variables:
+
+.. variable:: ExternalData_CUSTOM_LOCATION
+
+  When a custom fetch script is loaded, this variable is set to the
+  location part of the URL, which will contain the substituted hash
+  algorithm name and content hash value.
+
+.. variable:: ExternalData_CUSTOM_FILE
+
+  When a custom fetch script is loaded, this variable is set to the
+  full path to a file in which the script must store the fetched
+  content.  The name of the file is unspecified and should not be
+  interpreted in any way.
+
+The custom fetch script is expected to store fetched content in the
+file or set a variable:
+
+.. variable:: ExternalData_CUSTOM_ERROR
+
+  When a custom fetch script fails to fetch the requested content,
+  it must set this variable to a short one-line message describing
+  the reason for failure.
+
+#]=======================================================================]
+
+function(ExternalData_add_test target)
```

Comment:
Not strictly necessary, but it would be helpful to have a comment on each function summarizing intended behavior and detailing input formats and example usage. I.e. `# Usage: ExternalData_add_test(ReadAnImageData)`, etc.

Reply:
Yes, although we do not want to diverge from upstream *ExternalData.cmake*.

Reply:
> we do not want to diverge from upstream ExternalData.cmake

It doesn't seem like this constraint will be easy to enforce on a forked file here. Is there another way we could reference `ExternalData.cmake` from an existing ITK source dir or fetch it from an ITK hash with `FetchContent` at build time?

Reply:
We could require a newer version of CMake which does have IPFS support in ExternalData. Matt, do you know which is the oldest version good enough for this?

Reply:
To clarify, the IPFS support has not yet been contributed to upstream CMake [^1]

Waiting we integrate it upstream, an intermediate approach would be to maintain the modified ExternalData module into a project called for example `cmake-IPFS-ExternalData` maintained in the [@InsightSoftwareConsortium](https://github.com/InsightSoftwareConsortium) GitHub organization.

Structure would be similar to what we are doing for [scikit-build/cmake-FindVcvars](https://github.com/scikit-build/cmake-FindVcvars) and [scikit-build/cmake-FindFortran](https://github.com/scikit-build/cmake-FindFortran).

The proposal structure provide clear guideline for:
* integrating the custom module in projects
* versioning the module for easy integration

[^1]: https://github.com/Kitware/CMake/blob/master/Modules/ExternalData.cmake

Reply:
Yes, good idea. I will start a new dedicated repository for the module when integrating into the *ITK* repository.

Note that it intentionally has not been submitted to upstream CMake because of the current lack of availability of a C++ CID verification library.